### PR TITLE
feat(privacy): add user data export and account deletion workflow [AICC-74]

### PR DESCRIPTION
**Pull request by altosun154**
- Adds `GET /api/privacy/export` — returns a JSON file with the user's profile, favourites, comments, ratings, passkey metadata, and newsletter subscription (no password hashes or raw keys)
- Adds `POST /api/privacy/request-deletion` — issues a short-lived token (1 h TTL) to authorise account deletion
- Adds `DELETE /api/privacy/account` — verifies the token then hard-deletes the user and all associated records (comments, ratings, passkeys, newsletter subscription)
- Adds `/account/privacy` frontend page with a data download button and a 2-step account deletion flow (request → type DELETE → confirm)
- Wires "Privacy & Data" into the AuthDropdown menu for logged-in users

## Description

Please include a summary of the change and which issue is fixed.  
Reference any related issues using `#1234` notation.

**What is the problem this PR is solving?**

**How did you solve it?**

## Related Issue

Add Jira issue link here.

## Type of Change

Please delete options that are not relevant.

- 🐛 Bug fix (non-breaking change which fixes an issue)
- ✨ New feature (non-breaking change which adds functionality)
- 🔄 Refactor (non-breaking change that improves code structure/quality)
- 📖 Documentation update
- 🚀 Performance improvement
- ✅ Test addition or enhancement
- 🧹 Chore (build process, CI, tooling, etc)

## Which Workspace(s) Does This Affect?

_Select all that apply_:

- [ ] `frontend`
- [ ] `backend`
- [ ] `crawler`
- [ ] `bin` / CLI
- [ ] `newsletters`
- [ ] `.github` / GitHub Actions
- [ ] `shell` / scripts
- [ ] Root / tooling
- [ ] Other (please specify)

## Checklist

Please verify the following before requesting review:

- [ ] I have read the **contributing guidelines**.
- [ ] My code follows the existing **style guide** (lint, formatting).
- [ ] I have added or updated **unit tests** for new functionality.
- [ ] I have added or updated **end-to-end tests** where applicable.
- [ ] I have added or updated **documentation** (in-code comments, README, PR template).
- [ ] I have run the full **test suite** locally (`npm test`, `npm run test:e2e`).
- [ ] I have tested my changes in **development** and **production** builds.
- [ ] Any dependent changes have been merged and published in downstream modules.

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce:

```bash
# Example:
npm ci
npm run build
# run unit tests
npm run test
# run frontend E2E
npm --workspace frontend run test:e2e
# run crawler smoke
npm --workspace crawler run crawl
```

- [ ] Unit tests
- [ ] Integration tests
- [ ] End-to-end tests
- [ ] Manual testing steps (please describe)

## Deployment Notes

- Does this require new environment variables?

  - [ ] Yes — documented below
  - [ ] No

If yes, list and describe:

```text
ENV_VAR_NAME=description
```

## Rollback Plan

If something goes wrong, what is the rollback strategy?

---

Thank you for your contribution! 🎉
Please ensure all checks pass before merging.

### CODE REVIEW

File: backend/src/controllers/privacy.controller.ts [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+      message:
+        "Deletion token issued. Confirm within 1 hour to permanently delete your account.",
+      deletionToken,
+      expiresAt: deletionTokenExpiry.toISOString(),
+    });
+  } catch (error) {
+    console.error("Error requesting account deletion:", error);
+    return res.status(500).json({ error: "Internal server error" });
+  }
+};
```
[the remark this answers is not in the document]
Comment (altosun154) [reply backend/src/controllers/privacy.controller.ts:158]: Acknowledged — the token is intentionally a UX confirmation gate rather than a cryptographic second factor. The two-step round-trip (request → confirm) does add a minor CSRF hurdle since both calls require the same live JWT, but you're correct that it isn't a true second factor.

The stronger approach would be password re-entry (for password-based accounts) or a passkey assertion (for passkey-only accounts). This is deferred as a follow-up since it requires handling both credential types and this dashboard serves a relatively low-risk internal audience. Tracked for a future hardening pass.

File: backend/src/controllers/privacy.controller.ts [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+
+    return res.json({
+      message:
+        "Deletion token issued. Confirm within 1 hour to permanently delete your account.",
+      deletionToken,
+      expiresAt: deletionTokenExpiry.toISOString(),
+    });
+  } catch (error) {
+    console.error("Error requesting account deletion:", error);
+    return res.status(500).json({ error: "Internal server error" });
```
Comment (hoangsonww) [inline backend/src/controllers/privacy.controller.ts:156]: Non-constant-time comparison — leaks timing. After moving to hashed token storage (per Gemini's comment above), compare with `crypto.timingSafeEqual(Buffer.from(user.deletionToken, 'hex'), Buffer.from(hashedSubmitted, 'hex'))` rather than `!==`. Both buffers must be the same length, so guard with a length check first.

File: backend/src/controllers/privacy.controller.ts [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+/**
+ * DELETE /api/privacy/account
+ * Verifies the deletion token then hard-deletes the user and all linked records:
+ * comments, ratings, passkeys, newsletter subscription.
+ * Body: { deletionToken: string }
+ */
+export const confirmDeleteAccount = async (req: Request, res: Response) => {
+  try {
+    const userId = (req as any).user.id;
+    const { deletionToken } = req.body;
```
Comment (hoangsonww) [inline backend/src/controllers/privacy.controller.ts:187]: Cascade misses **`ChatSession`** (`backend/src/models/chatSession.ts`) — user-attributable content that survives deletion. Add `ChatSession.deleteMany({ userId })` here. Also worth including chat sessions in the export payload (line ~62) since they're GDPR-relevant personal data.

Comment (hoangsonww) [inline backend/src/controllers/privacy.controller.ts:187]: Dangling vote references: `Comment.upvotes` / `Comment.downvotes` on **other users'** comments still contain this user's ObjectId after deletion. Add:

```ts
Comment.updateMany(
  {},
  { $pull: { upvotes: userId, downvotes: userId } },
).session(session),
```

inside the transaction (per Gemini's transaction suggestion).

File: backend/src/controllers/privacy.controller.ts [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    res.setHeader("Content-Type", "application/json");
+    res.setHeader(
+      "Content-Disposition",
+      `attachment; filename="synthoraai-export-${userId}.json"`,
+    );
+    return res.json(exportPayload);
+  } catch (error) {
+    console.error("Error exporting user data:", error);
+    return res.status(500).json({ error: "Internal server error" });
+  }
```
Comment (hoangsonww) [inline backend/src/controllers/privacy.controller.ts:105]: No rate limit on `request-deletion`. A caller with a valid JWT can spam this endpoint, repeatedly overwriting the deletion token / churning DB writes. Add a per-user limit (e.g. 3/hour) — `express-rate-limit` keyed on `req.user.id` is sufficient.

Comment (hoangsonww) [inline backend/src/controllers/privacy.controller.ts:105]: Token is returned directly in the response body rather than emailed. The two-step request→confirm flow is therefore a UX gate, not a security gate (a stolen JWT completes both calls).

You've already acknowledged the JWT-only concern in reply to Copilot and deferred password/passkey re-auth as follow-up. Even so, emailing the token (or a confirmation link) to the registered address is a cheap interim step that proves email control — much lower lift than full re-auth and meaningfully raises the bar.

File: backend/src/controllers/privacy.controller.ts [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+      })),
+      newsletterSubscription: newsletter
+        ? {
+            subscribedAt: newsletter.subscribedAt,
+            lastSentAt: newsletter.lastSentAt ?? null,
+          }
+        : null,
+    };
+
+    res.setHeader("Content-Type", "application/json");
```
Comment (hoangsonww) [inline backend/src/controllers/privacy.controller.ts:96]: If the user cancels in the UI (e.g. closes the page after `request-deletion`), `deletionToken` / `deletionTokenExpiry` linger in the DB until the 1h TTL. Either:

1. Add a `POST /api/privacy/cancel-deletion` that clears these fields, and call it from `handleCancelDeletion` in the frontend, or
2. Clear these fields on next successful login.

Minor data-hygiene issue, but cheap to fix.

File: backend/src/controllers/privacy.controller.ts [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+        .lean(),
+    ]);
+
+    const exportPayload = {
+      exportedAt: new Date().toISOString(),
+      profile: {
+        id: user._id,
+        email: user.email,
+        name: user.name ?? null,
+        isVerified: user.isVerified,
```
Comment (hoangsonww) [inline backend/src/controllers/privacy.controller.ts:50]: Export omits **upvotes/downvotes the user cast on other comments** — that's personal data under GDPR. Add a fetch over `Comment.find({ $or: [{ upvotes: userId }, { downvotes: userId }] }, { _id: 1, upvotes: 1, downvotes: 1 })` and serialize a `votes` array into the payload (just IDs + which side they voted).

File: backend/src/routes/privacy.routes.ts [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+
+/**
+ * @swagger
+ * /api/privacy/export:
+ *   get:
+ *     tags: [Privacy]
+ *     summary: Export all personal data for the authenticated user
+ *     description: >
+ *       Returns a JSON file attachment containing the user's profile, favourites,
+ *       comments, ratings, passkey metadata, and newsletter subscription.
```
Comment (hoangsonww) [inline backend/src/routes/privacy.routes.ts:42]: Verify `ApiKeyAuth` matches the security scheme name used by other route docs in this repo — some routes use `bearerAuth`. If the swagger config doesn't define an `ApiKeyAuth` scheme, the "Authorize" UI on these endpoints will be broken. Keep it consistent across `auth.routes.ts`, `passkey.routes.ts`, etc.

File: frontend/pages/account/privacy.tsx [hunk 8]
```diff
@@ -0,0 +1,309 @@
...
+type DeletionStep = "idle" | "confirming" | "requested" | "done";
+
+export default function PrivacyPage() {
+  const router = useRouter();
+  const [exportLoading, setExportLoading] = useState(false);
+  const [deletionStep, setDeletionStep] = useState<DeletionStep>("idle");
+  const [deletionToken, setDeletionToken] = useState("");
+  const [deletionExpiry, setDeletionExpiry] = useState("");
+  const [deleteLoading, setDeleteLoading] = useState(false);
+  const [confirmText, setConfirmText] = useState("");
```
Comment (hoangsonww) [inline frontend/pages/account/privacy.tsx:30]: Direct `localStorage.getItem("token")` bypasses the `getAuthToken()` helper used everywhere else in `frontend/services/api.ts`. Use the helper — it centralizes the storage key and handles SSR / edge cases. Also: this runs on every render; if the goal is only to redirect unauthenticated users, the `useEffect` with empty/router deps is fine, but read via `getAuthToken()`.

File: frontend/pages/account/privacy.tsx [hunk 9]
```diff
@@ -0,0 +1,309 @@
...
+  };
+
+  const handleConfirmDeletion = async () => {
+    if (confirmText !== "DELETE") return;
+    setDeleteLoading(true);
+    try {
+      await confirmDeleteAccount(deletionToken);
+      setDeletionStep("done");
+      clearAuthToken();
+      toast.success("Your account has been permanently deleted.");
```
Comment (hoangsonww) [inline frontend/pages/account/privacy.tsx:77]: `setTimeout` without cleanup — if the user navigates away before 2500ms elapses, this still fires `router.replace("/")` post-unmount. Stash in a ref and clear in an effect cleanup:

```tsx
const redirectTimer = useRef<ReturnType<typeof setTimeout> | null>(null);
// ...
redirectTimer.current = setTimeout(() => router.replace("/"), 2500);
// cleanup
useEffect(() => () => { if (redirectTimer.current) clearTimeout(redirectTimer.current); }, []);
```

Or just call `router.replace("/")` immediately and let the toast linger on its own timeline.

**Review by hoangsonww — CHANGES_REQUESTED** [review]
Reviewed PR — requesting changes. Please also see and address the existing **gemini-code-assist** and **Copilot** reviews already on this PR (hashed token storage, atomic transactions, defensive `res.json().catch(...)`, newsletter email normalization, and the JWT-only deletion concern). Comments below cover items those bots did **not** flag.

## Additional findings (not duplicated from bots)

### Backend
1. **Non-constant-time token compare** — even after switching to hashed tokens per Gemini's suggestion, use `crypto.timingSafeEqual` rather than `!==`.
2. **Dangling vote references** — `Comment.upvotes` / `Comment.downvotes` on *other users'* comments still contain the deleted user's ObjectId. Add `Comment.updateMany({}, { $pull: { upvotes: userId, downvotes: userId } })` to the cascade.
3. **`ChatSession` not cleaned up** — `backend/src/models/chatSession.ts` is user-attributable and not deleted. Also missing from the export payload — GDPR-relevant personal data.
4. **No rate limit on `request-deletion`** — caller can spam tokens, repeatedly overwriting prior tokens. Add per-user limit (e.g. 3/hour).
5. **No email-based confirmation** — token is returned in the HTTP response rather than emailed. The two-step flow is a UX gate, not a security gate. (Author already acknowledged the JWT-only concern in reply to Copilot; emailing the token is the cheap upgrade path even if password/passkey re-auth is deferred.)
6. **Stale token never cleared on cancel** — if the user cancels in the UI, `deletionToken` / `deletionTokenExpiry` linger in the DB until the 1h TTL. Add a `POST /api/privacy/cancel-deletion` (or clear on next successful login) to scrub them.
7. **Export omits user-cast votes** — upvotes/downvotes the user cast on other users' comments are personal data per GDPR and should be in the export.
8. **No tests** — per `CLAUDE.md`, backend changes should ship with Jest coverage. Add happy-path + expired-token + missing-token + token-mismatch tests for `confirmDeleteAccount`, and at minimum a smoke test for `exportUserData`.
9. **Swagger `ApiKeyAuth`** — verify this matches the security scheme name used by other route docs in this repo (some use `bearerAuth`). Inconsistency degrades the generated docs.

### Frontend
10. **Direct `localStorage.getItem("token")`** in `privacy.tsx` bypasses the existing `getAuthToken()` helper used everywhere else in `services/api.ts`. Use the helper for consistency and SSR safety.
11. **`setTimeout` without cleanup** — the 2500ms redirect after deletion can fire after unmount. Stash the timer in a ref and `clearTimeout` in a cleanup function, or call `router.replace` immediately and let the toast linger via its own lifecycle.

Once the above + the gemini/Copilot items are addressed, this is good to merge.

File: backend/src/controllers/privacy.controller.ts (hunk 3 above)
Comment (altosun154) [reply backend/src/controllers/privacy.controller.ts:187]: Thanks for flagging this — you're right that ChatSession is user-attributable content and should be covered.

Looking at the current schema in backend/src/models/chatSession.ts, sessions are article-scoped rather than user-scoped — IChatSession only has articleId and messages, with no userId field. The model also isn't imported in any controller at the moment, so deleteMany({ userId }) as written would silently match nothing.

Happy to go either direction:

Defer to a follow-up — add a userId field to ChatSession + wire it in the chat controller as a separate PR, then the deletion and export can follow naturally.
Do it here — I can add userId to the schema, backfill the chat controller to populate it, and then include the deletion and export in this PR. That's a bit more scope but keeps the privacy coverage complete.
Which would you prefer? If (2), I'll keep it in one commit so the schema change, controller wiring, and privacy endpoints move together.

**Review by hoangsonww — APPROVED** [review]
